helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/StickyRebalanceStrategy.java

```diff
@@ -0,0 +1,205 @@
+      }
+    }
+    for (Map.Entry<String, Map<String, String>> entry : stateMap.entrySet()) {
+      znRecord.setListField(entry.getKey(), new ArrayList<>(entry.getValue().keySet()));
+    }
+    if (logger.isDebugEnabled()) {
+      logger.debug("znRecord: {}", znRecord);
+    }
+
+    return znRecord;
+  }
+
+  private int countStateReplicas() {
+    int total = 0;
+    for (Integer count : _states.values()) {
+      total += count;
+    }
+    return total;
+  }
+
+  /**
+   * Populates a valid state map from the current mapping, filtering out invalid nodes.
+   *
+   * @param currentMapping   the current mapping of partitions to node states
+   * @param assignableNodes  the list of nodes that can be assigned
+   * @return a map of partitions to valid node states
+   */
+  private Map<String, Map<String, String>> populateValidStateMapFromCurrentMapping(
+      final Map<String, Map<String, String>> currentMapping,
+      final List<CapacityNode> assignableNodes) {
+    Map<String, Map<String, String>> validStateMap = new HashMap<>();
+    // Convert the assignableNodes to map for quick lookup
+    Map<String, CapacityNode> assignableNodeMap =
+        assignableNodes.stream().collect(Collectors.toMap(CapacityNode::getId, node -> node));
+    if (currentMapping != null) {
+      for (Map.Entry<String, Map<String, String>> entry : currentMapping.entrySet()) {
+        String partition = entry.getKey();
+        Map<String, String> currentNodeStateMap = new HashMap<>(entry.getValue());
+        // Skip if current node state is invalid with state model
+        if (!isValidStateMap(currentNodeStateMap)) {
+          continue;
+        }
+        // Filter out invalid node assignment
+        currentNodeStateMap.entrySet()
+            .removeIf(e -> !isValidNodeAssignment(partition, e.getKey(), assignableNodeMap));
+
+        validStateMap.put(partition, currentNodeStateMap);
+      }
+    }
+    return validStateMap;
+  }
+
+  /**
+   * Validates whether the provided state mapping is valid according to the defined state model.
+   *
+   * @param currentNodeStateMap A map representing the actual state mapping where the key is the node ID and the value is the state.
+   * @return true if the state map is valid, false otherwise
+   */
+  private boolean isValidStateMap(final Map<String, String> currentNodeStateMap) {
+    // Check if the size of the current state map exceeds the total state count in state model
+    if (currentNodeStateMap.size() > countStateReplicas()) {
+      return false;
+    }
+
+    Map<String, Integer> tmpStates = new HashMap<>(_states);
+    for (String state : currentNodeStateMap.values()) {
+      // If the state is not defined in the state model, return invalid
+      if (!tmpStates.containsKey(state)) {
+        return false;
+      }
+      tmpStates.put(state, tmpStates.get(state) - 1);
+    }
+
+    // Ensure no state has a negative count
+    return tmpStates.values().stream().noneMatch(count -> count < 0);
```

Comment:
This is really expensive... Especially in multiple for loop. We should rely on previous checking code to validate it instead of this last line.